python/pyarrow/array.pxi

```diff
@@ -1543,7 +1543,20 @@ cdef class Array(_PandasConvertible):
     def _to_pandas(self, options, types_mapper=None, **kwargs):
         return _array_like_to_pandas(self, options, types_mapper=types_mapper)
 
-    def __array__(self, dtype=None):
+    def __array__(self, dtype=None, copy=None):
+        # TODO honor the copy=True case
```

Comment:
Can we raise an exception for now? Also, can you open a GH issue for it?

Reply:
> Can we raise an exception for now?

I wouldn't do that I think, because then when numpy would start passing that down in let's say numpy 2.1, `np.array(obj)` would start erroring, while those never errored before (and for some cases this might incorrectly not return a copy (although still marked as read-only), many cases already do copy anyways)

Although to be honest, I don't really know what the strategy of numpy will be to enable this keyword. If they would just start to stop copying the result of `__array__`, that would cause such changes in many libraries with `__array__`s)

Reply:
Though ideally, we just directly implement `copy=True` as well (it was just lest critical for numpy 2.0 as it is not yet used, and also a bit harder to implement)

Reply:
> and also a bit harder to implement

On second thought, it might be relatively easy to determine if `to_numpy()` returned a copy or not? I was first thinking we would have to mimic the logic based on the type ("if primitive and no nulls, then it is zero copy"), but we might be able to check if the resulting ndarray has a base pointing to a pyarrow object?

Although the simple logic of numeric+no nulls might be easier in practice.